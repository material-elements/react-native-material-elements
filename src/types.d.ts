export * from './components/types';
export * from './types/index';
