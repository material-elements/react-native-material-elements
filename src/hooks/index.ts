export * from './useRestyle';
export * from './useThemedProps';
