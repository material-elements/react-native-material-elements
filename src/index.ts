export * from './components';
export * from './hooks';
export * from './libraries';
export * from './utils';
