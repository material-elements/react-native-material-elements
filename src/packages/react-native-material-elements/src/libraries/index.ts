export * from './themes';
export * from './style';
