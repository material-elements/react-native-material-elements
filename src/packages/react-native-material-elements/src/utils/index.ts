export * from './styleGenerator';
export * from './utils';
export * from './dimensions';
