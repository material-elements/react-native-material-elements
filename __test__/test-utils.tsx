import React from 'react';
import { render, RenderOptions } from '@testing-library/react-native';
import { ThemeProvider } from '../src';

export const ThemeWrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  return <ThemeProvider>{children}</ThemeProvider>;
};

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: ThemeWrapper, ...options });

export * from '@testing-library/react-native';
export { customRender as render };
