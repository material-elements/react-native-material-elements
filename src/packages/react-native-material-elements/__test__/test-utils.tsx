import React from 'react';
import { render, RenderOptions } from '@testing-library/react-native';
import { ThemeProvider } from '../src';
import { ThemeProviderProps } from '../src/libraries/types';

interface CustomRenderOptions extends Omit<RenderOptions, 'wrapper'> {
  themeProps?: Partial<ThemeProviderProps>;
}

export const ThemeWrapper: React.FC<ThemeProviderProps> = ({ children, ...props }) => {
  return <ThemeProvider {...props}>{children}</ThemeProvider>;
};

export const customRender = (ui: React.ReactElement, options?: CustomRenderOptions) => {
  const { themeProps, ...rest } = options || {};
  return render(ui, {
    wrapper: wrapperProps => <ThemeWrapper {...themeProps}>{wrapperProps.children}</ThemeWrapper>,
    ...rest,
  });
};

export * from '@testing-library/react-native';
export { customRender as render };
