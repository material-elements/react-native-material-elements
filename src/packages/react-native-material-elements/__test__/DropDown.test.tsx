import * as RN from 'react-native';
import { DropDown, DropDownListContainer, gray, Text } from '../src';
import { fireEvent, render, waitFor } from './test-utils';

const mockData = [
  { id: 'bd7acbea-c1b1-46c2-aed5-3ad53abb28ba', title: 'First Item' },
  { id: '3ac68afc-c605-48d3-a4f8-fbd91aa97f63', title: 'Second Item' },
  { id: '58694a0f-3da1-471f-bd96-145571e29d72', title: 'Third Item' },
  { id: '58694a0f-3da1-471f-bd96-145571e29d12', title: 'Four Item' },
  { id: '58694a0f-3da1-471f-bd96-145571e29415', title: 'Five Item' },
  { id: '58694a0f-3da1-471f-bd96-145571e29419', title: 'Six Item' },
  { id: '58694a0f-3da1-471f-bd96-145571e29412', title: 'Seven Item' },
];

describe('DropDown Component', () => {
  const mockInputWrapperTouchTestId = 'input-wrapper-touch-test-id';
  const mockInputTestId = 'input-test-id';

  beforeEach(() => {
    jest.useFakeTimers();
    jest.clearAllMocks();
  });

  afterEach(() => {
    jest.clearAllTimers();
  });

  it('should render correctly', async () => {
    const { toJSON } = render(<DropDown data={mockData} />);
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it('should render the inputStartAdornment component correctly', () => {
    const label = 'inputStartAdornmentComponent';
    const { getByText } = render(<DropDown inputStartAdornment={<Text>{label}</Text>} data={mockData} />);
    const inputStartAdornment = getByText(label);
    expect(inputStartAdornment).toBeDefined();
  });

  it('should render the inputEndAdornment component correctly', () => {
    const label = 'inputEndAdornmentComponent';
    const { getByText } = render(<DropDown inputEndAdornment={<Text>{label}</Text>} data={mockData} />);
    const inputEndAdornment = getByText(label);
    expect(inputEndAdornment).toBeDefined();
  });

  it('should trigger the onDropDownClicked when clicking on the input', () => {
    const mockOnDropDownClicked = jest.fn();

    const { getByTestId } = render(
      <DropDown
        data={mockData}
        inputWrapperTouchableOpacityTestId={mockInputWrapperTouchTestId}
        onDropDownClicked={mockOnDropDownClicked}
      />,
    );

    const wrapper = getByTestId(mockInputWrapperTouchTestId);
    fireEvent.press(wrapper);
    expect(mockOnDropDownClicked).toHaveBeenCalled();
    expect(mockOnDropDownClicked).toHaveBeenCalledTimes(1);
  });

  it('should render the input component', () => {
    const { getByTestId } = render(<DropDown data={mockData} dropDownInputTestId={mockInputTestId} />);

    const input = getByTestId(mockInputTestId);
    expect(input).toBeDefined();
  });

  it('should render icon input component when variation prop passed as icon', () => {
    const { getByTestId } = render(<DropDown data={mockData} variation="icon" dropDownInputTestId={mockInputTestId} />);

    const input = getByTestId(mockInputTestId);
    expect(input).toBeDefined();
  });

  it('should show empty value when input component mount', () => {
    const { getByTestId } = render(<DropDown data={mockData} variation="icon" dropDownInputTestId={mockInputTestId} />);

    const input = getByTestId(mockInputTestId);
    expect(input.props.value).toEqual('');
  });

  it('should not render any input component when invalid variation prop passed', () => {
    const { queryByTestId } = render(
      <DropDown data={mockData} variation={'unknown' as any} dropDownInputTestId={mockInputTestId} />,
    );

    const input = queryByTestId(mockInputTestId);
    expect(input).toBeNull();
  });

  it('should show the selected list item title', () => {
    const { getByTestId } = render(
      <DropDown
        data={[{ id: '1', title: 'first_item' }]}
        selectedListItems={[{ id: '1' }]}
        variation="icon"
        dropDownInputTestId={mockInputTestId}
      />,
    );

    const input = getByTestId(mockInputTestId);

    expect(input.props.value).toEqual('first_item');
  });

  it('should show the multiselect message', () => {
    const { getByTestId } = render(
      <DropDown
        data={[{ id: '1', title: 'first_item' }]}
        selectedListItems={[{ id: '1' }]}
        variation="icon"
        multiselect
        dropDownInputTestId={mockInputTestId}
      />,
    );

    const input = getByTestId(mockInputTestId);

    expect(input.props.value).toEqual('Selected items 1');
  });
});

describe('DropDownListContainer component', () => {
  const mockListItemTestId = 'mock-list-item-test-id';

  const mockOnItemClickedHandler = jest.fn();
  const mockOnCloseHandler = jest.fn();

  beforeEach(() => {
    jest.useFakeTimers();
    jest.clearAllMocks();
  });

  afterEach(() => {
    jest.clearAllTimers();
  });

  it('should render properly with default props', () => {
    const { toJSON } = render(
      <DropDownListContainer
        inputLayoutRectangle={{ x: 0, y: 0, width: 0, height: 0 }}
        dropDownContainerRect={{ x: 0, y: 0, width: 0, height: 0, pageX: 0, pageY: 0 }}
        data={[]}
      />,
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('should render the list item', () => {
    const { getByTestId } = render(
      <DropDownListContainer
        inputLayoutRectangle={{ x: 0, y: 0, width: 0, height: 0 }}
        dropDownContainerRect={{ x: 0, y: 0, width: 0, height: 0, pageX: 0, pageY: 0 }}
        data={[{ id: '1', title: 'first_item' }]}
        listItemTestId={mockListItemTestId}
      />,
    );

    const listItem = getByTestId(mockListItemTestId);
    expect(listItem).toBeDefined();
  });

  it('should called the onItemClicked when onItemClicked props will passed and user tap on the list item', () => {
    const { getByTestId } = render(
      <DropDownListContainer
        inputLayoutRectangle={{ x: 0, y: 0, width: 0, height: 0 }}
        dropDownContainerRect={{ x: 0, y: 0, width: 0, height: 0, pageX: 0, pageY: 0 }}
        data={[{ id: '1', title: 'first_item' }]}
        onItemClicked={mockOnItemClickedHandler}
        listItemTestId={mockListItemTestId}
      />,
    );

    const listItem = getByTestId(mockListItemTestId);

    fireEvent(listItem, 'press', { nativeEvent: {} });

    expect(mockOnItemClickedHandler).toHaveBeenCalledTimes(1);
    expect(mockOnItemClickedHandler).toHaveBeenCalledWith({ id: '1', title: 'first_item' });
  });

  it('should show the gray[900] color for title text', () => {
    const { getByText } = render(
      <DropDownListContainer
        inputLayoutRectangle={{ x: 0, y: 0, width: 0, height: 0 }}
        dropDownContainerRect={{ x: 0, y: 0, width: 0, height: 0, pageX: 0, pageY: 0 }}
        data={[{ id: '1', title: 'first_item' }]}
        onItemClicked={mockOnItemClickedHandler}
        listItemTestId={mockListItemTestId}
      />,
    );

    const title = getByText('first_item');
    expect(title).toBeDefined();
    expect(title.props.style.color).toEqual(gray[900]);
  });

  it('should show the gray[50] color for title text if item is selected', () => {
    const { getByText } = render(
      <DropDownListContainer
        inputLayoutRectangle={{ x: 0, y: 0, width: 0, height: 0 }}
        dropDownContainerRect={{ x: 0, y: 0, width: 0, height: 0, pageX: 0, pageY: 0 }}
        data={[{ id: '1', title: 'first_item' }]}
        onItemClicked={mockOnItemClickedHandler}
        listItemTestId={mockListItemTestId}
        selectedListItems={[{ id: '1' }]}
        showSelectedItem
      />,
    );

    const title = getByText('first_item');
    expect(title).toBeDefined();
    expect(title.props.style.color).toEqual(gray[50]);
  });

  it('should show the light color of the title text when color scheme is dark', () => {
    jest.spyOn(RN, 'useColorScheme').mockReturnValue('dark');

    const { getByText } = render(
      <DropDownListContainer
        inputLayoutRectangle={{ x: 0, y: 0, width: 0, height: 0 }}
        dropDownContainerRect={{ x: 0, y: 0, width: 0, height: 0, pageX: 0, pageY: 0 }}
        data={[{ id: '1', title: 'first_item' }]}
        onItemClicked={mockOnItemClickedHandler}
        listItemTestId={mockListItemTestId}
      />,
    );

    const title = getByText('first_item');
    expect(title).toBeDefined();
    expect(title.props.style.color).toEqual(gray[50]);
  });

  it('should show the listItemEndAdornment item when list items is isSelected', () => {
    const { getByText } = render(
      <DropDownListContainer
        inputLayoutRectangle={{ x: 0, y: 0, width: 0, height: 0 }}
        dropDownContainerRect={{ x: 0, y: 0, width: 0, height: 0, pageX: 0, pageY: 0 }}
        data={[{ id: '1', title: 'first_item' }]}
        selectedListItems={[{ id: '1' }]}
        listItemEndAdornment={<Text>Hello</Text>}
        displaySelectedAdornment
        showSelectedItem
      />,
    );

    const listItemEndAdornment = getByText('Hello');
    expect(listItemEndAdornment).toBeDefined();
  });

  it('should render the search bar component', () => {
    const { toJSON, getByPlaceholderText } = render(
      <DropDownListContainer
        inputLayoutRectangle={{ x: 0, y: 0, width: 0, height: 0 }}
        dropDownContainerRect={{ x: 0, y: 0, width: 0, height: 0, pageX: 0, pageY: 0 }}
        search
        data={mockData}
      />,
    );

    expect(toJSON()).toMatchSnapshot();

    const search = getByPlaceholderText('Search');
    expect(search).toBeDefined();
  });

  it('should call the onClose function when press on item', () => {
    const { getByText } = render(
      <DropDownListContainer
        inputLayoutRectangle={{ x: 0, y: 0, width: 0, height: 0 }}
        dropDownContainerRect={{ x: 0, y: 0, width: 0, height: 0, pageX: 0, pageY: 0 }}
        data={[{ id: '1', title: 'first_item' }]}
        onItemClicked={mockOnItemClickedHandler}
        listItemTestId={mockListItemTestId}
        onClose={mockOnCloseHandler}
      />,
    );

    const title = getByText('first_item');
    fireEvent(title, 'press', { nativeEvent: {} });
    expect(mockOnCloseHandler).toHaveBeenCalled();
    expect(mockOnCloseHandler).toHaveBeenCalledTimes(1);
  });
});
