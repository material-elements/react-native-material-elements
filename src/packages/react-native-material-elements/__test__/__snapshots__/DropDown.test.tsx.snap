// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown Component should render correctly 1`] = `
<View
  collapsable={false}
  style={
    [
      undefined,
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderColor": "#BDBDBD",
          "borderRadius": 5,
          "borderWidth": 0.6,
          "display": "flex",
          "flexDirection": "row",
          "opacity": 1,
          "paddingHorizontal": 14,
          "position": "relative",
          "width": "100%",
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "#ffffff",
            "left": 8,
            "paddingHorizontal": 8,
            "position": "absolute",
            "top": "50%",
            "transform": [
              {
                "translateY": -5,
              },
              {
                "scale": 1,
              },
            ],
            "zIndex": 10,
          }
        }
      >
        <Text
          collapsable={false}
          onLayout={[Function]}
          style={
            {
              "color": "#424242",
              "fontSize": 14,
              "opacity": 1,
            }
          }
        >
          Drop down
        </Text>
      </View>
      <TextInput
        editable={false}
        onBlur={[Function]}
        onFocus={[Function]}
        onLayout={[Function]}
        pointerEvents="none"
        style={
          {
            "backgroundColor": "transparent",
            "color": "#424242",
            "flex": 1,
            "minHeight": 35,
            "position": "relative",
            "width": "100%",
            "zIndex": 12,
          }
        }
        value=""
      />
    </View>
  </View>
</View>
`;

exports[`DropDownListContainer component should render properly with default props 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  transparent={true}
  visible={true}
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {},
        {},
        [
          {
            "backgroundColor": "rgba(0,0,0,0.5)",
            "flex": 1,
          },
          [
            {
              "backgroundColor": "transparent",
              "flex": 1,
              "paddingHorizontal": 10,
            },
          ],
        ],
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessible={false}
      collapsable={false}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#E0E0E0",
          "borderRadius": 5,
          "marginTop": 5,
          "maxHeight": undefined,
          "overflow": "hidden",
          "top": 0,
          "width": "100%",
        }
      }
    >
      <RCTScrollView
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        keyboardShouldPersistTaps="handled"
        nestedScrollEnabled={true}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        overScrollMode="always"
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEnabled={true}
        scrollEventThrottle={0.0001}
        stickyHeaderIndices={[]}
        style={
          [
            {
              "flexGrow": 1,
              "paddingVertical": 2,
            },
          ]
        }
        viewabilityConfigCallbackPairs={[]}
      >
        <View />
      </RCTScrollView>
    </View>
  </View>
</Modal>
`;

exports[`DropDownListContainer component should render the search bar component 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  transparent={true}
  visible={true}
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {},
        {},
        [
          {
            "backgroundColor": "rgba(0,0,0,0.5)",
            "flex": 1,
          },
          [
            {
              "backgroundColor": "transparent",
              "flex": 1,
              "paddingHorizontal": 10,
            },
          ],
        ],
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessible={false}
      collapsable={false}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#E0E0E0",
          "borderRadius": 5,
          "marginTop": 5,
          "maxHeight": undefined,
          "overflow": "hidden",
          "top": 0,
          "width": "100%",
        }
      }
    >
      <View
        style={
          [
            {
              "paddingHorizontal": 5,
              "paddingVertical": 4,
            },
            {},
            undefined,
          ]
        }
      >
        <View>
          <View
            style={
              [
                {
                  "backgroundColor": "#ECEFF1",
                },
                {
                  "testID": "undefined-wrapper",
                },
                [
                  {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                    "paddingHorizontal": 10,
                    "width": "100%",
                  },
                  undefined,
                  {
                    "borderColor": "#757575",
                    "borderWidth": 0.7,
                    "height": 30,
                  },
                ],
              ]
            }
            testID="undefined-wrapper"
          >
            <TextInput
              onChangeText={[Function]}
              placeholder="Search"
              placeholderTextColor="#757575"
              style={
                {
                  "color": "#212121",
                  "flex": 1,
                  "minHeight": 35,
                  "width": "100%",
                }
              }
            />
          </View>
        </View>
      </View>
      <RCTScrollView
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        keyboardShouldPersistTaps="handled"
        nestedScrollEnabled={true}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        overScrollMode="always"
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEnabled={true}
        scrollEventThrottle={0.0001}
        stickyHeaderIndices={[]}
        style={
          [
            {
              "flexGrow": 1,
              "paddingVertical": 2,
            },
          ]
        }
        viewabilityConfigCallbackPairs={[]}
      >
        <View />
      </RCTScrollView>
    </View>
  </View>
</Modal>
`;
